libs/core/futures/src/future_data.cpp

```diff
@@ -131,6 +131,11 @@ namespace hpx::lcos::detail {
         // this thread would block on the future
         threads::thread_id_ref_type runs_child = runs_child_;
 
+        if (!runs_child)
+        {
+            return false;
+        }
+
```

Comment:
Good catch, that also mean we can remove these lines: https://github.com/STEllAR-GROUP/hpx/blob/3a5a47a0a5fcc853971742bf853b92eb16239ded/libs/core/futures/src/future_data.cpp#L120-L123

Reply:
Sure, just left them in the case that taking a thread_id_ref_type was expensive. Give a thumbs up if you think it's ok to change it

Reply:
I have looked again and I have convinced myself that there is no need to protect the `runs_child_` intrusive pointer needs to be protected in any way. The issue has to be somewhere else... we need to keep looking.

Reply:
Ok, I didn't understand exactly where the issue was supposed to occur, so I'll just take your word for it.

Currently I can barely reproduce the current assert failure (where a tid from terminated_threads_ does not appear in the thread_map_, when it should), and the issue goes away when logging... I'll give it another go, might find a more reliable failure case

Reply:
I think this assert is wrong to begin with, I'm looking at it as we speak.